### PR TITLE
Testing build fix

### LINKED ISSUE #1981: [Flake] Failed to update config to git component deployed
[kind/bug]

## What versions of software are you using?
- Operating System: OpenShift CI
- Output of `odo version`: pr run

## How did you run odo exactly?
Run pr (on two test node) in OpenShift CI. I have seen this even on single test node.  

## Actual behavior
Flakey

## Expected behavior
Should not be flakey 

## Any logs, error output, etc?

Reference  - https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_odo/1920/pull-ci-openshift-odo-master-integration/1126

### PR DESCRIPTION
**Pull request by mik-dass**
**What kind of PR is this?**

/kind flake

**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1981 

**How to test changes / Special notes to the reviewer**:

### CODE REVIEW

**Comment by mik-dass** [comment]
```
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=backend-app'
```
/retest

**Comment by mik-dass** [comment]
```
Container setup exited with code 1, reason Error
```
/retest

**Comment by mik-dass** [comment]
```
error: could not run steps: step integration-e2e-benchmark failed: template pod "integration-e2e-benchmark" failed: the pod ci-op-hgs9sfdi/integration-e2e-benchmark failed after 44m39s (failed containers: setup, test): ContainerFailed one or more containers exited
```
/retest

**Comment by mik-dass** [comment]
```
error: could not run steps: step integration-e2e-benchmark failed: template pod "integration-e2e-benchmark" failed: the pod ci-op-3q88t7ms/integration-e2e-benchmark failed after 11m46s (failed containers: setup, test): ContainerFailed one or more containers exited
```
/retest

File: pkg/occlient/occlient.go [hunk 1]
```diff
@@ -1682,23 +1683,36 @@ func (c *Client) WaitForBuildToFinish(buildName string) error {
 		return errors.Wrapf(err, "unable to watch build")
 	}
 	defer w.Stop()
+	timeout := time.After(5 * time.Minute)
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: Shouldn't this be a constant? (high above in `occlient.go`)

File: pkg/occlient/occlient.go [hunk 2]
```diff
@@ -1672,7 +1672,8 @@ func (c *Client) StartBuild(name string) (string, error) {
 }
 
 // WaitForBuildToFinish block and waits for build to finish. Returns error if build failed or was canceled.
-func (c *Client) WaitForBuildToFinish(buildName string) error {
+func (c *Client) WaitForBuildToFinish(buildName string, stdout io.Writer) error {
+	following := false
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: Need more description on this, bit confused why we're setting following to false then true when going through the channel loop.

File: pkg/occlient/occlient.go [hunk 3]
```diff
@@ -1682,23 +1683,36 @@ func (c *Client) WaitForBuildToFinish(buildName string) error {
...
+					return errors.Errorf("build %s status %s", e.Name, e.Status.Phase)
+				case buildv1.BuildPhaseRunning:
+					// since the pod is ready and the build is now running, start following the logs
+					if !following {
+						following = true
+						err := c.FollowBuildLog(buildName, stdout)
+						if err != nil {
+							return err
+						}
+					}
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: All of this above looks good!

File: pkg/occlient/occlient.go [hunk 4]
```diff
@@ -1845,6 +1859,7 @@ func (c *Client) FollowBuildLog(buildName string, stdout io.Writer) error {
 	}
 
 	rd, err := c.buildClient.RESTClient().Get().
+		Timeout(5*time.Minute).
```
Comment (cdrage) [inline pkg/occlient/occlient.go]: Should be a constant

**Comment by mik-dass** [comment]
@cdrage Fixed

File: pkg/occlient/occlient.go [hunk 5]
```diff
@@ -1672,7 +1673,9 @@ func (c *Client) StartBuild(name string) (string, error) {
 }
 
 // WaitForBuildToFinish block and waits for build to finish. Returns error if build failed or was canceled.
-func (c *Client) WaitForBuildToFinish(buildName string) error {
+func (c *Client) WaitForBuildToFinish(buildName string, stdout io.Writer) error {
+	// following indicates if we have already setup the following logic
+	following := false
 	glog.V(4).Infof("Waiting for %s  build to finish", buildName)
 
 	w, err := c.buildClient.Builds(c.Namespace).Watch(metav1.ListOptions{
```
Comment (dharmit) [inline pkg/occlient/occlient.go]: I understand this is not touched in the proposed PR but can we add a comment that describes what this all about? At least I didn't understand a thing here. :disappointed:

File: pkg/occlient/occlient.go [hunk 6]
```diff
@@ -1682,23 +1685,37 @@ func (c *Client) WaitForBuildToFinish(buildName string) error {
 		return errors.Wrapf(err, "unable to watch build")
 	}
 	defer w.Stop()
+	timeout := time.After(OcBuildTimeout)
 	for {
```
Comment (dharmit) [inline pkg/occlient/occlient.go]: Don't understand what's going on in this `for` loop either. :disappointed:

**Comment by mik-dass** [comment]
@dharmit Updated with more comments

**Comment by mik-dass** [comment]
```
[odo]  ✗  invalid configuration: [context was not found for specified context: tmeibqyxnu/api-ci-op-t9t1htld-00a90-origin-ci-int-aws-dev-rhcloud-com:6443/developer, cluster has no server defined]
[odo] Please login to your server: 
```
/retest

File: pkg/occlient/occlient.go (hunk 6 above)
Comment (girishramnani) [reply pkg/occlient/occlient.go]: I can help here a bit. So here first thing to observe is the `select` statement on a quick TLDR: select statements are like channel collectors - listen on multiple channels at once, more here https://tour.golang.org/concurrency/5. 

now the `Build().Watch()` returns a channel which returns messages on as per the pod status. Now the change that has been done here is 
we wait for the pod to start before we start following the logs.

Comment (girishramnani) [reply pkg/occlient/occlient.go]: as for the need for the `following`, The `w.ResultChan()` can send multiple messages of type `BuildPhaseRunning`. But we dont have to follow logs multiple times, just the first time should be good enough. Hence this flag.
Think of it as a singleton

**Pull request merged**